qiskit/providers/experiment/experiment_data.py

```diff
@@ -0,0 +1,674 @@
+
+        job_ids = job_ids or []
+        self._jobs = OrderedDict((k, None) for k in job_ids)
+        self._job_futures = []
+        self._errors = []
+
+        self._data_queue = queue.Queue()
+        self._data = []
+
+        figure_names = figure_names or []
+        self._figures = dict.fromkeys(figure_names)
+        self._figures_queue = queue.Queue()
+        self._figure_names = figure_names  # Only used by add_figure.
+
+        self._analysis_results_queue = queue.Queue()
+        self._analysis_results = []
+
+        self._created_in_db = False
+
+    def add_data(
+            self,
+            data: Union[Result, List[Result], Job, List[Job], Dict, List[Dict]],
+            post_processing_callback: Optional[Callable] = None
+    ):
+        """Add experiment data.
+
+        Args:
+            data: Experiment data to add.
+                Several types are accepted for convenience:
+
+                    * Result: Add data from this ``Result`` object.
+                    * List[Result]: Add data from the ``Result`` objects.
+                    * Job: Add data from the job result.
+                    * List[Job]: Add data from the job results.
+                    * Dict: Add this data.
+                    * List[Dict]: Add this list of data.
+            post_processing_callback: Callback function invoked when all pending
+                jobs finish. This ``ExperimentData`` object is the only argument
+                to be passed to the callback function.
+        """
+        if isinstance(data, dict):
+            self._add_single_data(data)
+        elif isinstance(data, Result):
+            self._add_result_data(data)
+        elif isinstance(data, (Job, BaseJob)):
+            self._jobs[data.job_id()] = data
+            self._job_futures.append(
+                (data, self._executor.submit(self._wait_for_job, data, post_processing_callback)))
+            if self.auto_save:
+                self.save()
+        elif isinstance(data, list):
+            for dat in data:
+                self.add_data(dat)
+        else:
+            raise QiskitError(f"Invalid data type {type(data)}.")
+
+    def _wait_for_job(
+            self, job: Union[Job, BaseJob],
+            job_done_callback: Optional[Callable] = None
+    ) -> None:
+        """Wait for a job to finish.
+
+        Args:
+            job: Job to wait for.
+            job_done_callback: Callback function to invoke when job finishes.
+        """
+        LOG.debug(f"Waiting for job {job.job_id()} to finish.")
+        self._add_result_data(job.result())
+        if job_done_callback:
+            job_done_callback(self)
+
+    def _add_result_data(self, result: Result) -> None:
+        """Add data from a Result object
+
+        Args:
+            result: Result object containing data to be added.
+        """
+        for i in range(len(result.results)):
+            data = result.data(i)
+            data['job_id'] = result.job_id
```

Comment:
The design doc mentions circuit level metadata, which is indeed passed in the current `ExperumentData` placeholder class in the `qiskit-experiments` repository. The `ExperimentDataV1` class here does not store circuit metadata after all (unlike what's written in the design doc), is it intentional?

Reply:
Could you then please add it:
```
data['metadata'] = result.results[i].header.metadata
```

Reply:
Yeah it was added [here](https://github.com/Qiskit/qiskit-terra/blob/254b3cb121b9b83a06e7efee0781d555cba5c399/qiskit/providers/experiment/experiment_data.py#L233)